### ISSUE TITLE
weave-net CrashLoopBackOff for the second node

### ISSUE THREAD

**Issue by avkonst** [kubernetes/kubernetes#34101]
**Is this a request for help?**

I think it is an issue either with software or documentation, but I am not quite sure. I have started with a question on stackoverflow: http://stackoverflow.com/questions/39872332/how-to-fix-weave-net-crashloopbackoff-for-the-second-node

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

I think it is a bug or request to improve documentation

**Kubernetes version** (use `kubectl version`):
1.4.0

**Environment**:
- **Cloud provider or hardware configuration**: Vagrant
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04
- **Kernel** (e.g. `uname -a`):
- **Install tools**: kubeadm init/join
- **Others**:

**What happened**:

I have got 2 VMs nodes. Both see each other either by hostname (through /etc/hosts) or by ip address. One has been provisioned with kubeadm as a master. Another as a worker node. Following the instructions (http://kubernetes.io/docs/getting-started-guides/kubeadm/) I have added weave-net. The list of pods looks like the following:

```
vagrant@vm-master:~$ kubectl get pods --all-namespaces
NAMESPACE     NAME                                    READY     STATUS             RESTARTS   AGE
kube-system   etcd-vm-master                          1/1       Running            0          3m
kube-system   kube-apiserver-vm-master                1/1       Running            0          5m
kube-system   kube-controller-manager-vm-master       1/1       Running            0          4m
kube-system   kube-discovery-982812725-x2j8y          1/1       Running            0          4m
kube-system   kube-dns-2247936740-5pu0l               3/3       Running            0          4m
kube-system   kube-proxy-amd64-ail86                  1/1       Running            0          4m
kube-system   kube-proxy-amd64-oxxnc                  1/1       Running            0          2m
kube-system   kube-scheduler-vm-master                1/1       Running            0          4m
kube-system   kubernetes-dashboard-1655269645-0swts   1/1       Running            0          4m
kube-system   weave-net-7euqt                         2/2       Running            0          4m
kube-system   weave-net-baao6                         1/2       CrashLoopBackOff   2          2m

```

CrashLoopBackOff appears for each worker node connected. I have spent several ours playing with network interfaces, but it seems the network is fine. I have found similar question on stackoverflow, where the answer advised to look into the logs and no follow up. So, here are the logs:

```
vagrant@vm-master:~$ kubectl logs weave-net-baao6 -c weave --namespace=kube-system
2016-10-05 10:48:01.350290 I | error contacting APIServer: Get https://100.64.0.1:443/api/v1/nodes: dial tcp 100.64.0.1:443: getsockopt: connection refused; trying with blank env vars
2016-10-05 10:48:01.351122 I | error contacting APIServer: Get http://localhost:8080/api: dial tcp [::1]:8080: getsockopt: connection refused
Failed to get peers

```

**What you expected to happen**:

I would expect the weave-net to be in Running state

**How to reproduce it** (as minimally and precisely as possible):

I have not done anything special, just followed the documentation on Getting Started. If it is essencial, I can share Vagrant project, which I used to provision everything. Please, let me know if you need one.

**Comment by avkonst** [comment]
Please, let me know if it needs to be reported on weave-net project.

**Comment by bach1** [comment]
I reported a similar issue here:
http://stackoverflow.com/questions/39869583/how-to-get-kube-dns-working-in-vagrant-cluster-using-kubeadm-and-weave

**Comment by avkonst** [comment]
I wonder if there is any working step by step instruction on how to get kubernetes cluster up and running on a set of virtual machines? I am happy to downgrade the version of kubernetes if it is an option. I am totally confused with all of these options: kube-deploy, kube-up, and other, I found from 3rd parties..

**Comment by andreagardiman** [comment]
I have the same issue. I'm using VirtualBox to run 2 VM based on minimal Centos 7 image. All VMs are attached to 2 interfaces, a NAT and an host-only network.
The two VMs are able to connect to each other using the host-only network interfaces.

I tried also with the instructions about Calico and Canal, and I cannot make them work either.

**Comment by avkonst** [comment]
It seems Calico has got similar issue:

```
vagrant@vm-master:~$ kubectl get pods --all-namespaces
NAMESPACE     NAME                                    READY     STATUS             RESTARTS   AGE
kube-system   calico-etcd-iemkd                       1/1       Running            0          57m
kube-system   calico-node-178gc                       1/2       CrashLoopBackOff   8          56m
kube-system   calico-node-zsym2                       2/2       Running            0          57m
kube-system   calico-policy-controller-6gh7b          1/1       Running            0          57m
kube-system   etcd-vm-master                          1/1       Running            0          56m
kube-system   kube-apiserver-vm-master                1/1       Running            0          58m
kube-system   kube-controller-manager-vm-master       1/1       Running            0          57m
kube-system   kube-discovery-982812725-7jsmb          1/1       Running            0          57m
kube-system   kube-dns-2247936740-xiee1               3/3       Running            0          57m
kube-system   kube-proxy-amd64-iywb7                  1/1       Running            0          57m
kube-system   kube-proxy-amd64-ok9bx                  1/1       Running            0          56m
kube-system   kube-scheduler-vm-master                1/1       Running            0          56m
kube-system   kubernetes-dashboard-1655269645-g4cyd   1/1       Running            0          57m
  info: 1 completed object(s) was(were) not shown in pods list. Pass --show-all to see all objects.

vagrant@vm-master:~$ kubectl logs calico-node-178gc -c calico-node --namespace=kube-system
Waiting for etcd connection...
No IP provided. Using detected IP: 10.0.10.11
Traceback (most recent call last):
  File "startup.py", line 336, in <module>
    main()
  File "startup.py", line 287, in main
    warn_if_hostname_conflict(ip)
  File "startup.py", line 210, in warn_if_hostname_conflict
    current_ipv4, _ = client.get_host_bgp_ips(hostname)
  File "/usr/lib/python2.7/site-packages/pycalico/datastore.py", line 134, in wrapped
    "running?" % (fn.__name__, e.message))
pycalico.datastore_errors.DataStoreError: get_host_bgp_ips: Error accessing etcd (Connection to etcd failed due to MaxRetryError("HTTPConnectionPool(host='100.78.232.136', port=6666): Max retries exceeded with url: /v2/keys/calico/bgp/v1/host/vm-worker/ip_addr_v4 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f50416bfc50>, 'Connection to 100.78.232.136 timed out. (connect timeout=60)'))",)).  Is etcd running?
Calico node failed to start
```
What could I try to progress this issue further?

**Comment by tedstirm** [comment]
Okay I ran into this exact same issue and here is how I fixed it. 

This problem seems to be due to kube-proxy looking at the wrong network interface. If you look at the kube-proxy logs on a worker node you will most likely see something like:
```
-A KUBE-SEP-4C6YEJQ2VXV53FEZ -m comment --comment default/kubernetes:https -s 10.0.2.15/32 -j KUBE-MARK-MASQ
```
This is the wrong network interface. The kube-proxy should be looking at the master node's IP address not the NAT IP address. 

As far as I know the kube-proxy gets this value from the Kube API Server when starting up. If you look at the Kube API Server's documentation it states that if `--advertise-address` flag isn't set it will default to `--bind-address` and if `--bind-address` isn't set it will default to host's default interface. Which in my case and yours seems to be the NAT interface, which isn't what we want. So what I did was set the Kube API Server's `--advertise-address` flag and everything started working. So right after Step 2 and before Step 3 of 
> [Installing Kubernetes on Linux with kubeadm](http://kubernetes.io/docs/getting-started-guides/kubeadm/)

You will need to update your `/etc/kubernetes/manifests/kube-apiserver.json` and add the `--advertise-address` flag to point to your master node's IP address.

So for example: My master node's IP address is `172.28.128.2`, which means I update my `/etc/kubernetes/manifests/kube-apiserver.json` to look like:
```
cat <<EOF > /etc/kubernetes/manifests/kube-apiserver.json
{
  "kind": "Pod",
  "apiVersion": "v1",
  "metadata": {
    "name": "kube-apiserver",
    "namespace": "kube-system",
    "creationTimestamp": null,
    "labels": {
      "component": "kube-apiserver",
      "tier": "control-plane"
    }
  },
  "spec": {
    "volumes": [
      {
        "name": "certs",
        "hostPath": {
          "path": "/etc/ssl/certs"
        }
      },
      {
        "name": "pki",
        "hostPath": {
          "path": "/etc/kubernetes"
        }
      }
    ],
    "containers": [
      {
        "name": "kube-apiserver",
        "image": "gcr.io/google_containers/kube-apiserver-amd64:v1.4.0",
        "command": [
          "/usr/local/bin/kube-apiserver",
          "--v=4",
          "--insecure-bind-address=127.0.0.1",
          "--etcd-servers=http://127.0.0.1:2379",
          "--admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota",
          "--service-cluster-ip-range=100.64.0.0/12",
          "--service-account-key-file=/etc/kubernetes/pki/apiserver-key.pem",
          "--client-ca-file=/etc/kubernetes/pki/ca.pem",
          "--tls-cert-file=/etc/kubernetes/pki/apiserver.pem",
          "--tls-private-key-file=/etc/kubernetes/pki/apiserver-key.pem",
          "--token-auth-file=/etc/kubernetes/pki/tokens.csv",
          "--secure-port=443",
          "--allow-privileged",
          "--advertise-address=172.28.128.2",
          "--etcd-servers=http://127.0.0.1:2379"
        ],
        "resources": {
          "requests": {
            "cpu": "250m"
          }
        },
        "volumeMounts": [
          {
            "name": "certs",
            "mountPath": "/etc/ssl/certs"
          },
          {
            "name": "pki",
            "readOnly": true,
            "mountPath": "/etc/kubernetes/"
          }
        ],
        "livenessProbe": {
          "httpGet": {
            "path": "/healthz",
            "port": 8080,
            "host": "127.0.0.1"
          },
          "initialDelaySeconds": 15,
          "timeoutSeconds": 15
        }
      }
    ],
    "hostNetwork": true
  },
  "status": {}
}
EOF
```

I am not to sure about the listed above as a long term solution as what happens if the <kube-component/etcd>.json changes? By kube component I mean: kube-apiserver, kube-controller-manage, kube-scheduler, etcd. Ideally, I think the user would want just some way to set these flags via kubeadm or maybe the user should be responsible and just parse the JSON themselves, I'm not sure. Thoughts?

However, it still may be a good idea to update Step 2 of [Installing Kubernetes on Linux with kubeadm](http://kubernetes.io/docs/getting-started-guides/kubeadm/)  to atleast mention to the users that they can update the kube component flags by modifying the their json found at: `/etc/kubernetes/manifests/`.

**Comment by errordeveloper** [comment]
I am looking at this right now. We have a way of reproducing it with https://github.com/davidkbainbridge/k8s-playground, where we have already applied a fix that makes kubelet pickup desride IP for pods in host network namespaces, however somehow `kubernetes` service IP still points to 10.0.2.15.

**Comment by chenzhiwei** [comment]
I encountered this issue too.

Adding --advertise-address when starting kube-apiserver solved this issue.

**Comment by errordeveloper** [comment]
@chenzhiwei I've already started diving into the code, and you saved me time to realise how this part worked exactly, thank you so much! This looks like an easy fix for kubeadm, PR is coming shortly.

**Comment by errordeveloper** [comment]
Ok, so it turns out that this flag is not enough, we still have an issue reaching `kubernetes` service IP. The simplest solution to this is to run `kube-proxy` with `--proxy-mode=userspace`. To enable this, you can use `kubectl -n kube-system edit ds kube-proxy-amd64 && sudo kubectl -n kube-system delete pods -l name=kube-proxy-amd64`.

**Comment by avkonst** [comment]
I am starting kubeadm with --advertise-address option:
kubeadm init --api-advertise-addresses=$master_address
where my $master_address is not NAT interface. It still not enough to remove this issue.

**Comment by avkonst** [comment]
"Ok, so it turns out that this flag is not enough, we still have an issue reaching kubernetes service IP. The simplest solution to this is to run kube-proxy with --proxy-mode=userspace. To enable this, you can use kubectl -n kube-system edit ds kube-proxy-amd64 && kubectl -n kube-system delete pods -l name=kube-proxy-amd64."

@errordeveloper Could you, please, explain what flag are you setting, to what command in the getting started tutorial? 

The last command you wrote, what does it do?

**Comment by errordeveloper** [comment]
@avkonst see https://github.com/kubernetes/kubernetes/pull/34607.

Also, you can do this for now:
```
jq \
   '.spec.containers[0].command |= .+ ["--advertise-address=172.42.42.1"]' \
   /etc/kubernetes/manifests/kube-scheduler.json > /tmp/kube-scheduler.json
mv /tmp/kube-scheduler.json /etc/kubernetes/manifests/kube-scheduler.json
```

**Comment by errordeveloper** [comment]
> Could you, please, explain what flag are you setting, to what command in the getting started tutorial?

We are going to fix this shortly. If you can hop on Slack, I'd be happy to help.

**Comment by errordeveloper** [comment]
> What does your last command do?

As this thread is getting quite noisy, here is a recap.

First, find out what IP address you want to use on the master, it's probably the one on the second network interface. For this example I'll use `IP="172.42.42.1"`.

Next, run `kubeadm init --api-advertise-addresses=$IP`.

Now, you want to append `--advertise-address` to `kube-apiserver` in the static pod manifest, you can do it like this:
```
jq \
   '.spec.containers[0].command |= .+ ["--advertise-address=$IP"]' \
   /etc/kubernetes/manifests/kube-scheduler.json > /tmp/kube-scheduler.json
mv /tmp/kube-scheduler.json /etc/kubernetes/manifests/kube-scheduler.json
```

And finally, you need to update flags in `kube-proxy` daemonset and append `--proxy-mode=userspace`, which can be done like this:
```
kubectl -n kube-system get ds -l 'component=kube-proxy-amd64' -o json \
  | jq '.items[0].spec.template.spec.containers[0].command |= .+ ["--proxy-mode=userspace"]' \
  |   kubectl apply -f - && kubectl -n kube-system delete pods -l 'component=kube-proxy-amd64'
```

**Comment by errordeveloper** [comment]
I'm working on docs for this. Also, @davidkbainbridge has provided a Vagrant+Ansible implementation, and latest fixes are in [my fork](https://github.com/errordeveloper/kubernetes-vagrant-ansible-ubuntu).

**Comment by miry** [comment]
@tedstirm @errordeveloper  can you help me with similar problem, I have multiple eth*. And I am using `10.136.0.0/16` for private  network, but `eth1`. For now I fixed issue with weave via open port 443 for eth0 and eth1. But found next issue:

I could not access to the instances from the cluster(example to redis that hosts on separate instance). I found that weave uses `10.0.0.0/8` network. And I am looking for solution to change this. Do you know how can I change it? Example to `101.2.0.0/16`?

**Comment by tedstirm** [comment]
@errordeveloper I didn't have to update the `--proxy-mode` for my kube-proxy.  All I had to do was make sure the `--advertise-address` flag for the Kube API Server was set to a value listed in the `--api-advertise-addresses` ? I'm just curious as to why you need to set `--proxy-mode` in your setup and I don't?  Any ideas? I looked over the [fork](https://github.com/errordeveloper/kubernetes-vagrant-ansible-ubuntu) and your Vagrant setup looks very similar to mine, except you are using Ubuntu vs my Centos.

**Comment by errordeveloper** [comment]
@miry 
> I found that weave uses `10.0.0.0/8` network.

No, it uses [`10.32.0.0/12`](https://github.com/weaveworks/weave-kube/blob/1be4819fdb01b2774144d12cb42bcd2f554e9e5c/image/launch.sh#L6).

So for Weave we have `10.32.0.0/12`, which expands to `10.32.0.1` - `10.47.255.254`.
Kubernetes service range is `10.12.0.0/12`, which expands to `10.0.0.1` - `10.15.255.254`.
And you have `10.136.0.0/16`, which expands to `10.136.0.1` - `10.136.255.254`.
I don't see an overlap here, which is what you seems to suggest, but please correct me if I missunderstood. I think you problem is different, and please file another issue. Your previous issue #34031 didn't not describe the problem, please try to describe the problem.

@tedstirm 
> ...your Vagrant setup looks very similar to mine, except you are using Ubuntu vs my Centos.

How many VMs are you running on? If it's a single-node setup, you won't notice anything. If it's not, I'm quite curious and would like to take a closer look, could you share your Vagrantfile?

**Comment by geotro** [comment]
@errordeveloper, after applying your changes I was able to get kube-dns and weave-net working!

However, kube-scheduler-kubernetes won't run:

````
# kubectl logs kube-scheduler-kubernetes --namespace=kube-system
unknown flag: --advertise-address
Usage of /usr/local/bin/kube-scheduler:
      --address string                           The IP address to serve on (set to 0.0.0.0 for all interfaces) (default "0.0.0.0")
      --algorithm-provider string                The scheduling algorithm provider to use, one of: DefaultProvider | ClusterAutoscalerProvider (default "DefaultProvider")
      --alsologtostderr value                    log to standard error as well as files
      --failure-domains string                   Indicate the "all topologies" set for an empty topologyKey when it's used for PreferredDuringScheduling pod anti-affinity. (default "kubernetes.io/hostname,failure-domain.beta.kubernetes.io/zone,failure-domain.beta.kubernetes.io/region")
      --feature-gates value                      A set of key=value pairs that describe feature gates for alpha/experimental features. Options are:
...

**Comment by errordeveloper** [comment]
@geotro sorry, that was a copy/paste bug, I've updated the comments now, it's should had been `kube-apiserver.json` not `kube-scheduler.json`.

**Comment by geotro** [comment]
Yeah I figured that out shortly after my comment.  I have everything working as expected now.  Thank you to everyone who participated in fixing this! :)

**Comment by errordeveloper** [comment]
@geotro you are welcome! We should eliminate the need for these work-around soon.

**Comment by bach1** [comment]
Thank you for the feedback. I'm getting closer to a working solution. With my basic Vagrant 2-node cluster (Centos) I now have weave up and running. The kube-dns service is however still not working.
Steps taken:
1) kubeadm init on master with api-advertise-addresses flag set.
2) Update flags in pod manifest and daemonsets per guidelines above.
3) Join node.
4) Install weave on master.

kube-dns is still stuck in ContainerCreating state:
```
NAMESPACE     NAME                             READY     STATUS              RESTARTS   AGE
kube-system   etcd-master                      1/1       Running             0          25m
kube-system   kube-apiserver-master            1/1       Running             1          17m
kube-system   kube-controller-manager-master   1/1       Running             0          20m
kube-system   kube-discovery-982812725-zlcw3   1/1       Running             0          20m
kube-system   kube-dns-2247936740-thz2p        0/3       ContainerCreating   0          19m
kube-system   kube-proxy-amd64-79prj           1/1       Running             0          17m
kube-system   kube-proxy-amd64-qroee           1/1       Running             0          16m
kube-system   kube-scheduler-master            1/1       Running             0          25m
kube-system   weave-net-35udy                  2/2       Running             0          15m
kube-system   weave-net-hs2xa                  2/2       Running             0          15m
```
I experimented with both bridged network settings and private network settings but the problem still persists.

**Comment by errordeveloper** [comment]
@bach1 could you provide the output of `kubectl -n kube-system describe pod -l name=kube-dns`?

**Comment by miry** [comment]
@errordeveloper 

> Kubernetes service range is 10.12.0.0/12, which expands to 10.0.0.1 - 10.15.255.254.

I thought it should be `100.64.0.0/12`. 

After update api server to use specific ip address the wave could not connect to 100.64.0.1:443.

Digital Ocean has also Anchor Ip to eth0  with mask: 10.10.0.0/16:

```
$ ip route | head 5
default via 159.203.176.1 dev eth0
10.10.0.0/16 dev eth0  proto kernel  scope link  src 10.10.0.12 
10.32.0.0/12 dev weave  proto kernel  scope link  src 10.32.0.1 
10.136.0.0/16 dev eth1  proto kernel  scope link  src 10.136.26.23 
159.203.176.0/20 dev eth0  proto kernel  scope link  src 159.203.188.178
```

**Comment by bach1** [comment]
@errordeveloper sure, see below. I assume the last cni-related errors propagate before weave is installed. The local NAT-interface address is still visible in `Node: master/10.0.2.15`. Is this an issue?
```
Name:		kube-dns-2247936740-yi89m
Namespace:	kube-system
Node:		master/10.0.2.15
Start Time:	Thu, 13 Oct 2016 09:46:44 +0000
Labels:		component=kube-dns
		k8s-app=kube-dns
		kubernetes.io/cluster-service=true
		name=kube-dns
		pod-template-hash=2247936740
		tier=node
Status:		Pending
IP:		
Controllers:	ReplicaSet/kube-dns-2247936740
Containers:
  kube-dns:
    Container ID:	
    Image:		gcr.io/google_containers/kubedns-amd64:1.7
    Image ID:		
    Ports:		10053/UDP, 10053/TCP
    Args:
      --domain=cluster.local
      --dns-port=10053
    Limits:
      cpu:	100m
      memory:	170Mi
    Requests:
      cpu:		100m
      memory:		170Mi
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Liveness:		http-get http://:8080/healthz delay=60s timeout=5s period=10s #success=1 #failure=1
    Readiness:		http-get http://:8081/readiness delay=30s timeout=5s period=10s #success=1 #failure=3
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cx6kh (ro)
    Environment Variables:	<none>
  dnsmasq:
    Container ID:	
    Image:		gcr.io/google_containers/kube-dnsmasq-amd64:1.3
    Image ID:		
    Ports:		53/UDP, 53/TCP
    Args:
      --cache-size=1000
      --no-resolv
      --server=127.0.0.1#10053
    Limits:
      cpu:	100m
      memory:	170Mi
    Requests:
      cpu:		100m
      memory:		170Mi
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cx6kh (ro)
    Environment Variables:	<none>
  healthz:
    Container ID:	
    Image:		gcr.io/google_containers/exechealthz-amd64:1.1
    Image ID:		
    Port:		8080/TCP
    Args:
      -cmd=nslookup kubernetes.default.svc.cluster.local 127.0.0.1:53 >/dev/null && nslookup kubernetes.default.svc.cluster.local 127.0.0.1:10053 >/dev/null
      -port=8080
      -quiet
    Limits:
      cpu:	10m
      memory:	50Mi
    Requests:
      cpu:		10m
      memory:		50Mi
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cx6kh (ro)
    Environment Variables:	<none>
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	False 
  PodScheduled 	True 
Volumes:
  default-token-cx6kh:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-cx6kh
QoS Class:	Guaranteed
Tolerations:	dedicated=master:NoSchedule
Events:
  FirstSeen	LastSeen	Count	From			SubobjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  8m		8m		1	{default-scheduler }			Normal		Scheduled	Successfully assigned kube-dns-2247936740-yi89m to master
  5m		5m		1	{kubelet master}			Warning		FailedMount	MountVolume.SetUp failed for volume "kubernetes.io/secret/f347bff6-9129-11e6-9232-525400c583ad-default-token-cx6kh" (spec.Name: "default-token-cx6kh") pod "f347bff6-9129-11e6-9232-525400c583ad" (UID: "f347bff6-9129-11e6-9232-525400c583ad") with: Get https://172.42.42.1:443/api/v1/namespaces/kube-system/secrets/default-token-cx6kh: dial tcp 172.42.42.1:443: getsockopt: connection refused
  8m		3m		130	{kubelet master}			Warning		FailedSync	Error syncing pod, skipping: failed to "SetupNetwork" for "kube-dns-2247936740-yi89m_kube-system" with SetupNetworkError: "Failed to setup network for pod \"kube-dns-2247936740-yi89m_kube-system(f347bff6-9129-11e6-9232-525400c583ad)\" using network plugins \"cni\": cni config unintialized; Skipping pod"
```

**Comment by errordeveloper** [comment]
@miry 
> I thought it should be `100.64.0.0/12`.

You a right, it's that in the current release, but it changed in master.

> After update api server to use specific ip address the wave could not connect to 100.64.0.1:443.
> Digital Ocean has also Anchor Ip to eth0 with mask: 10.10.0.0/16...

Yes, there is an issue with that in some DO regions... you need to pass `--service-cidr`, but not with the build you are using.

If you are online now, it'd be easy if you could find me on Slack. Thanks.

**Comment by errordeveloper** [comment]
@bach1 just as a sanity check, could you kill the DNS pod, i.e. `kubectl -n kube-system delete pod -l name=kube-dns --all`, may there is some bug on the CNI end and somehow it's not getting picked up for this pod...

**Comment by bach1** [comment]
@errordeveloper Hmm no luck. After deletion:
```
[root@master ~]get pods --all-namespacese=kube-dns
NAMESPACE     NAME                             READY     STATUS              RESTARTS   AGE
kube-system   etcd-master                      1/1       Running             0          3h
kube-system   kube-apiserver-master            1/1       Running             1          3h
kube-system   kube-controller-manager-master   1/1       Running             0          3h
kube-system   kube-discovery-982812725-qpndv   1/1       Running             0          3h
kube-system   kube-dns-2247936740-iffl7        0/3       ContainerCreating   0          2h
kube-system   kube-dns-2247936740-yi89m        0/3       Terminating         0          3h
kube-system   kube-proxy-amd64-7jgeg           1/1       Running             0          3h
kube-system   kube-proxy-amd64-e73eu           1/1       Running             0          3h
kube-system   kube-scheduler-master            1/1       Running             0          3h
kube-system   weave-net-lnvyx                  2/2       Running             0          3h
kube-system   weave-net-oox1r                  2/2       Running             0          3h
[root@master ~]# kubectl -n kube-system describe pod -l name=kube-dns
Name:		kube-dns-2247936740-iffl7
Namespace:	kube-system
Node:		node-1/10.0.2.15
Start Time:	Thu, 13 Oct 2016 11:11:37 +0000
Labels:		component=kube-dns
		k8s-app=kube-dns
		kubernetes.io/cluster-service=true
		name=kube-dns
		pod-template-hash=2247936740
		tier=node
Status:		Pending
IP:		
Controllers:	ReplicaSet/kube-dns-2247936740
Containers:
  kube-dns:
    Container ID:	
    Image:		gcr.io/google_containers/kubedns-amd64:1.7
    Image ID:		
    Ports:		10053/UDP, 10053/TCP
    Args:
      --domain=cluster.local
      --dns-port=10053
    Limits:
      cpu:	100m
      memory:	170Mi
    Requests:
      cpu:		100m
      memory:		170Mi
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Liveness:		http-get http://:8080/healthz delay=60s timeout=5s period=10s #success=1 #failure=1
    Readiness:		http-get http://:8081/readiness delay=30s timeout=5s period=10s #success=1 #failure=3
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cx6kh (ro)
    Environment Variables:	<none>
  dnsmasq:
    Container ID:	
    Image:		gcr.io/google_containers/kube-dnsmasq-amd64:1.3
    Image ID:		
    Ports:		53/UDP, 53/TCP
    Args:
      --cache-size=1000
      --no-resolv
      --server=127.0.0.1#10053
    Limits:
      cpu:	100m
      memory:	170Mi
    Requests:
      cpu:		100m
      memory:		170Mi
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cx6kh (ro)
    Environment Variables:	<none>
  healthz:
    Container ID:	
    Image:		gcr.io/google_containers/exechealthz-amd64:1.1
    Image ID:		
    Port:		8080/TCP
    Args:
      -cmd=nslookup kubernetes.default.svc.cluster.local 127.0.0.1:53 >/dev/null && nslookup kubernetes.default.svc.cluster.local 127.0.0.1:10053 >/dev/null
      -port=8080
      -quiet
    Limits:
      cpu:	10m
      memory:	50Mi
    Requests:
      cpu:		10m
      memory:		50Mi
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cx6kh (ro)
    Environment Variables:	<none>
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	False 
  PodScheduled 	True 
Volumes:
  default-token-cx6kh:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-cx6kh
QoS Class:	Guaranteed
Tolerations:	dedicated=master:NoSchedule
No events.

Name:				kube-dns-2247936740-yi89m
Namespace:			kube-system
Node:				master/10.0.2.15
Start Time:			Thu, 13 Oct 2016 09:46:44 +0000
Labels:				component=kube-dns
				k8s-app=kube-dns
				kubernetes.io/cluster-service=true
				name=kube-dns
				pod-template-hash=2247936740
				tier=node
Status:				Terminating (expires Thu, 13 Oct 2016 11:12:07 +0000)
Termination Grace Period:	30s
IP:				
Controllers:			ReplicaSet/kube-dns-2247936740
Containers:
  kube-dns:
    Container ID:	
    Image:		gcr.io/google_containers/kubedns-amd64:1.7
    Image ID:		
    Ports:		10053/UDP, 10053/TCP
    Args:
      --domain=cluster.local
      --dns-port=10053
    Limits:
      cpu:	100m
      memory:	170Mi
    Requests:
      cpu:		100m
      memory:		170Mi
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Liveness:		http-get http://:8080/healthz delay=60s timeout=5s period=10s #success=1 #failure=1
    Readiness:		http-get http://:8081/readiness delay=30s timeout=5s period=10s #success=1 #failure=3
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cx6kh (ro)
    Environment Variables:	<none>
  dnsmasq:
    Container ID:	
    Image:		gcr.io/google_containers/kube-dnsmasq-amd64:1.3
    Image ID:		
    Ports:		53/UDP, 53/TCP
    Args:
      --cache-size=1000
      --no-resolv
      --server=127.0.0.1#10053
    Limits:
      cpu:	100m
      memory:	170Mi
    Requests:
      cpu:		100m
      memory:		170Mi
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cx6kh (ro)
    Environment Variables:	<none>
  healthz:
    Container ID:	
    Image:		gcr.io/google_containers/exechealthz-amd64:1.1
    Image ID:		
    Port:		8080/TCP
    Args:
      -cmd=nslookup kubernetes.default.svc.cluster.local 127.0.0.1:53 >/dev/null && nslookup kubernetes.default.svc.cluster.local 127.0.0.1:10053 >/dev/null
      -port=8080
      -quiet
    Limits:
      cpu:	10m
      memory:	50Mi
    Requests:
      cpu:		10m
      memory:		50Mi
    State:		Waiting
      Reason:		ContainerCreating
    Ready:		False
    Restart Count:	0
    Volume Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-cx6kh (ro)
    Environment Variables:	<none>
Conditions:
  Type		Status
  Initialized 	True 
  Ready 	False 
  PodScheduled 	True 
Volumes:
  default-token-cx6kh:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-cx6kh
QoS Class:	Guaranteed
Tolerations:	dedicated=master:NoSchedule
No events.
```

**Comment by miry** [comment]
resolved my issue via add routing rule to use eth1 for node machines to kubernetes service range ip. Example:

```
echo "100.64.0.0/12 dev eth1" >> /etc/sysconfig/network-scripts/route-eth1
ip route add 100.64.0.0/12 dev eth1
```

**Comment by miry** [comment]
@petergardfjall but weave-net  was up 15m ago and dns last event was 25m ago. try to restart dns pod.

**Comment by errordeveloper** [comment]
It would be easier if folks could hop on #kubeadm channel in Slack and we
can debug step-by-step. It became really hard to pick individual setup
differences in this thread.

**Comment by petergardfjall** [comment]
@miry I did restart the pod with no luck (by the way, when you say _restart_ a pod, I assume you mean deleting the pod (and have the replica set replace it), right)?

It is stuck in `ContainerCreating`, although it appears to have reached slightly farther, beeing able to at least create containers. However 

```
# kubectl logs -n kube-system kube-dns-2247936740-f8nxa kube-dns
Error from server: Get https://kube-slave2:10250/containerLogs/kube-system/kube-dns-2247936740-f8nxa/kube-dns: dial tcp: lookup kube-slave2 on 10.0.2.3:53: no such host
```

But I guess I'll follow @errordeveloper's advice and take it on the slack channel.

**Comment by petergardfjall** [comment]
So I actually did manage to get it working at last.

The last missing piece of the puzzle was to follow the advice of point 5) in http://kubernetes.io/docs/getting-started-guides/kubeadm/#limitations:

> If you are using VirtualBox (directly or via Vagrant), you will need to ensure that hostname -i 
> returns a routable IP address (i.e. one on the second network interface, not the first one). 
> By default, it doesn’t do this and kubelet ends-up using first non-loopback network interface, which is usually NATed.

If anyone is interested, my `Vagrantfile` ended up looking as follows:

```
VAGRANTFILE_API_VERSION = "2"

Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
  config.vm.box = "ubuntu/xenial64"
  # configure a slightly larger than deafult (384MB) virtual box
  config.vm.provider "virtualbox" do |v|
    v.customize ["modifyvm", :id, "--memory", 1548] 
  end
  
  config.vm.define "kube-master", primary: true do |master|
    master.vm.hostname = "kube-master"
    master.vm.network :private_network, ip: "172.17.1.101"
    master.vm.network :forwarded_port, guest: 443, host: 8443
    # hostname -i must return a routable address on second (non-NATed) network interface
    # see 5) in http://kubernetes.io/docs/getting-started-guides/kubeadm/#limitations
    master.vm.provision :shell, :inline => "sed 's/127.0.0.1.*kube-master/172.17.1.101 kube-master/' -i /etc/hosts"
    master.vm.provision :shell, :path => "master-bootstrap.sh"    
  end

  config.vm.define "kube-slave1" do |slave|
    slave.vm.hostname = "kube-slave1"
    slave.vm.network :private_network, ip: "172.17.1.201"
    # hostname -i must return a routable address on second (non-NATed) network interface
    # see 5) in http://kubernetes.io/docs/getting-started-guides/kubeadm/#limitations    
    slave.vm.provision :shell, :inline => "sed 's/127.0.0.1.*kube-slave1/172.17.1.201 kube-slave1/' -i /etc/hosts"
    slave.vm.provision :shell, :path => "slave-bootstrap.sh"
  end

  config.vm.define "kube-slave2" do |slave|
    slave.vm.hostname = "kube-slave2"
    slave.vm.network :private_network, ip: "172.17.1.202"
    # hostname -i must return a routable address on second (non-NATed) network interface
    # see 5) in http://kubernetes.io/docs/getting-started-guides/kubeadm/#limitations
    slave.vm.provision :shell, :inline => "sed 's/127.0.0.1.*kube-slave2/172.17.1.202 kube-slave2/' -i /etc/hosts"
    slave.vm.provision :shell, :path => "slave-bootstrap.sh"
  end
  
end
```

**Comment by errordeveloper** [comment]
I am very happy to hear you got it working, we planning to add a proper
vagrant how-to, once we have tested latest packages. Have you referred to
the Ansible code in k8s-playgound repo linked from the docs?

**Comment by petergardfjall** [comment]
@errordeveloper: yes, I did have a look at that code and just adapted their `sed` snippet to my particular setup.

**Comment by bach1** [comment]
@petergardfjall : thanks for the hint. I can also confirm that I have a working system now.

**Comment by errordeveloper** [comment]
Have you enabled userspace mode in kube-proxy also?

**Comment by enieuw** [comment]
I run a similar setup based on CoreOS boxes(two nics, eth0 nat and eth1 private) on Vagrant. I had to make two changes to get the networking to work:
- Set the --node-ip on each kubelet process to the private network ip for each node
- Enable masquerading on the proxy to respond over the correct interface(--masquerade-all)

**Comment by errordeveloper** [comment]
Pease checkout https://github.com/errordeveloper/kubernetes-ansible-vagrant, there is code to handle firewalld rules and SELinux stuff (if you are on CentOS), as well as VirtualBox issues (applies to Ubuntu as well as CentOS). We are planing to provide plain text documentation soon too. Find me on Slack if you still have problems. It's always a good idea to create a new issue rather then comment on an existing one that seems like it describes your problem.

**Comment by raghu67** [comment]
I am seeing a similar issue. Here are the details:
[root@m0062421 ~]# kubectl version
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:16:57Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:10:32Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
[root@m0062421 ~]# kubeadm version
kubeadm version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-alpha.0.1534+cf7301f16c0363-dirty", GitCommit:"cf7301f16c036363c4fdcb5d4d0c867720214598", GitTreeState:"dirty", BuildDate:"2016-09-27T18:10:39Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
[root@m0062421 ~]# 

CentOS 7.2. Kernel Version:
[root@m0062421 ~]# uname -a
Linux m0062421.lab.ppops.net 3.10.0-327.18.2.el7.x86_64 #1 SMP Thu May 12 11:03:55 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

These are based on an old version of OpenStack and KVM. if that is relevant, I can find the details

**Comment by avkonst** [comment]
@errordeveloper, letting you know that workaround command is not being rejected:

```
ubuntu@k8s1:~$ kubectl -n kube-system get ds -l 'component=kube-proxy-amd64' -o json | jq '.items[0].spec.template.spec.containers[0].command |= .+ ["--cluster-cidr=10.32.0.0/12"]' | kubectl
apply --validate=false -f - && kubectl -n kube-system delete pods -l 'component=kube-proxy-amd64'
Object 'Kind' is missing in '{
      "spec": {
        "template": {
          "spec": {
            "containers": [
              {
                "command": [
                  "--cluster-cidr=10.32.0.0/12"
                ]
              }
            ]
          }
        }
      }
    }'
```

What could I do to fix it?

**Comment by jamstar** [comment]
i am having the same issue as @avkonst.   If you leave validation on, you get this error. `error validating "STDIN": error validating data: items[0].apiVersion not set;`
i included the input from `kubectl -n kube-system get ds -l 'component=kube-proxy-amd64' -o json | jq '.items[0].spec.template.spec.containers[0].command |= .+ ["--proxy-mode=userspace"]' ` and both kind and apiVersion are clearly there.  
`{
  "kind": "List",
  "apiVersion": "v1",
  "metadata": {},
  "items": [
    {
      "spec": {
        "template": {
          "spec": {
            "containers": [
              {
                "command": [
                  "--cluster-cidr=10.32.0.0/12"
                ]
              }
            ]
          }
        }
      }
    }
  ]
}
`

`kubectl  apply -f -` it seems to not get the outermost portion of the input which is where the kind and apiVersion are both at.

**Comment by thedos1701** [comment]
Hi, I can confirm the problem still exists on a two nodes kubernetes cluster installed with Kubeadm.
The nodes are Ubuntu Xenial 16.04 connected to each other using a tinc VPN.
Everything works fine on the tainted master node but on the second (worker only) node every CNI network plugin container keep crashing. (I tried Flannel and Weake)
It seems like the network-fabric container is unable to reach the API server on 16.96.0.1
I tried the workaround proposed but couldn't get it through becouse of the issue psoted by @jamstar and apparentely the first two steps are insufficient.

**Comment by miry** [comment]
@thedos1701 you need add route to 16.96.0.1 via your you VPN interface.

https://github.com/kubernetes/kubernetes/issues/34101#issuecomment-253745940

**Comment by tsvenkat** [comment]
@tedstirm
In my case, my apiserver's advertise-address was correct, but I still had the crashloop issue with the weave pods. Adding "--proxy-mode=userspace" using @errordeveloper 's neat jq trick solved the issue for me.

@errordeveloper
Though the issue is solved, I don't have a good understanding of why it solved. I looked at the kube-proxy documentation(http://kubernetes.io/docs/admin/kube-proxy) and it says:

> Which proxy mode to use: 'userspace' (older) or 'iptables' (faster). If blank, look at the Node object on the Kubernetes API and respect the 'net.experimental.kubernetes.io/proxy-mode' annotation if provided.  Otherwise use the best-available proxy (currently iptables).  If the iptables proxy is selected, regardless of how, but the system's kernel or iptables versions are insufficient, this always falls back to the userspace proxy.

Should we expect any side effects as we are setting it to the "older" option? Thanks.

**Comment by JulienLenne** [comment]
Hi everyone,
I just finished a Vagrant box with centos 7.2 and kubeadm with weave-net and dashboard.
[vagrant-centos-kubernetes](https://github.com/JulienLenne/vagrant-centos-kubernetes)
I include all solved issues in this post. Thank you to @petergardfjall and @errordeveloper !

@avkonst replace all occurences of kube-proxy-amd64 by kube-proxy maybe ... just check it with a get nodes

**Comment by mikedanese** [comment]
Moved to kubernetes/kubeadm#66

**Comment by thomasmodeneis** [comment]
I'm also experiencing the same issue:
```
kube-system   weave-net-0rlkr                        1/2       CrashLoopBackOff   10         30m
kube-system   weave-net-35tkz                        1/2       CrashLoopBackOff   10         30m
kube-system   weave-net-ph8wj                        1/2       CrashLoopBackOff   10         3m
```

Any ideas on how to deal with the issue ?

**Comment by marccarre** [comment]
@thomasmodeneis, what do the `weave` container's logs say?

**Comment by luxas** [comment]
@thomasmodeneis Please open a new issue in the kubeadm repo to track that

**Comment by thomasmodeneis** [comment]
@marccarre would you mind sending me the cmd for getting this logs out into a gist for example ? that would help. 
@luxas I will raise a new ticket asap.

**Comment by marccarre** [comment]
@thomasmodeneis either:
- `$ docker logs <weave-kube's container ID>` or
- `$ kubectl logs weave-net-0rlkr weave -n kube-system`
